SimG4CMS/HGCalTestBeam/plugins/HGCalTiminAnalyzer.cc

```diff
@@ -0,0 +1,350 @@
+  tok_hitsEE_  = consumes<edm::PCaloHitContainer>(edm::InputTag("g4SimHits",tmp1));
+#ifdef EDM_ML_DEBUG
+  std::cout << "HGCalTimingAnalyzer:: Detector " << detectorEE_ 
+	    << " with tags " << tmp1 << std::endl;
+#endif
+  tmp1         = iConfig.getParameter<std::string>("CaloHitSrcBeam");
+  tok_hitsBeam_= consumes<edm::PCaloHitContainer>(edm::InputTag("g4SimHits",tmp1));
+#ifdef EDM_ML_DEBUG
+  std::cout << "HGCalTimingAnalyzer:: Detector " << detectorBeam_ 
+	    << " with tags " << tmp1 << std::endl;
+#endif
+}
+
+HGCalTimingAnalyzer::~HGCalTimingAnalyzer() {}
+
+void HGCalTimingAnalyzer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  desc.add<std::string>("DetectorEE","HGCalEESensitive");
+  desc.add<std::string>("DetectorBeam","HcalTB06BeamDetector");
+  desc.add<bool>("GroupHits",false);
+  desc.add<double>("TimeUnit",0.001);
+  std::vector<int> ids = {1001,1002,1003,1004,1005};
+  desc.add<std::vector<int>>("IDBeams",ids);
+  desc.addUntracked<bool>("DoTree",true);
+  desc.add<edm::InputTag>("GeneratorSrc",edm::InputTag("generatorSmeared"));
+  desc.add<std::string>("CaloHitSrcEE","HGCHitsEE");
+  desc.add<std::string>("CaloHitSrcBeam","HcalTB06BeamHits");
+  descriptions.add("HGCalTimingAnalyzer",desc);
+}
+
+void HGCalTimingAnalyzer::beginJob() {
+  std::string det(detectorEE_); 
+  if (doTree_) {
+    tree_ = fs_->make<TTree>("HGCTB","SimHitEnergy");
+    tree_->Branch("xBeam",             &xBeam_,           "xBeam/D");
+    tree_->Branch("yBeam",             &yBeam_,           "yBeam/D");
+    tree_->Branch("zBeam",             &zBeam_,           "zBeam/D");
+    tree_->Branch("pBeam",             &pBeam_,           "pBeam/D");
+    tree_->Branch("simHitCellIdEE_",   &simHitCellIdEE_);
+    tree_->Branch("simHitCellEnEE_",   &simHitCellEnEE_);
+    tree_->Branch("simHitCellTmEE_",   &simHitCellTmEE_);
+    tree_->Branch("simHitCellIdBeam_", &simHitCellIdBeam_);
+    tree_->Branch("simHitCellEnBeam_", &simHitCellEnBeam_);
+    tree_->Branch("simHitCellTmBeam_", &simHitCellTmBeam_);
+  }
+}
+
+void HGCalTimingAnalyzer::beginRun(const edm::Run&, const edm::EventSetup& iSetup) {
+  edm::ESHandle<HGCalDDDConstants>  pHGDC;
+  iSetup.get<IdealGeometryRecord>().get(detectorEE_, pHGDC);
+  hgcons_ = &(*pHGDC);
+    
+#ifdef EDM_ML_DEBUG
+  std::cout << "HGCalTimingAnalyzer::" << detectorEE_ << " defined with "
+	    << hgcons_->layers(false) << " layers" << std::endl;
+#endif
+  
+}
+
+void HGCalTimingAnalyzer::analyze(const edm::Event& iEvent, 
+				  const edm::EventSetup& iSetup) {
+
+  //Generator input
+  edm::Handle<edm::HepMCProduct> evtMC;
+  iEvent.getByToken(tok_hepMC_,evtMC);
+  if (!evtMC.isValid()) {
+    edm::LogWarning("HGCal") << "no HepMCProduct found";
+  } else { 
+    const HepMC::GenEvent * myGenEvent = evtMC->GetEvent();
+    unsigned int k(0);
+    for (HepMC::GenEvent::particle_const_iterator p = myGenEvent->particles_begin();
```

Comment:
It is left for future development. We leave this code within the EDM_ML_DEBUG for the time being